Various changes needed in order to handle Dedicon test books

bertfrees opened the pull request:
Commit bb28b93 was your attempt to fix the oscillation issue (#70).
I believe commit b13a6dd fixes a bug in your commit.
Commit 80185f8 is to improve the speed at which the volume breaking algorithm converges to an optimum.
Commit b8dc003 was part of my initial attempt to fix the oscillation issue, but it's still relevant I think.
Commits b662b0d, b13a6dd, fe7804e and 051d099 were done by me and Paul when I was at Dedicon some weeks ago. After this the oscillation issue was fixed.
Commits e34b5dd and aa8bfe9 fix two other issues that I discovered during my visit to Dedicon. The first issue was that the braille page numbering would break at volume transitions in some cases. The number would become even on right hand pages where it was uneven in the previous volume, or visa versa. The second issue was that we would sometimes get a StackOverFlowError during the resolution of a marker-reference.

bertfrees committed:
Only save "NextPageDetailsInSequence" and "TransitionProperties" info when needed

bertfrees commented on src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:
```diff
@@ -191,7 +191,9 @@ private PageImpl nextPageInner(int pageNumberOffset, boolean hyphenateLastLine,
 		PageImpl current = newPage(pageNumberOffset);
 		if (pcbl!=null) {
-			blockContext.getRefs().setNextPageDetailsInSequence(pcbl, current.getDetails());
+			if (transitionContent.isPresent() && transitionContent.get().getType()==TransitionContent.Type.INTERRUPT) {
```
Only set if this is the last page of the volume (and there is a volume transition). `getNextPageDetailsInSequence` is only used in combination with `getTransitionProperties`, and `getTransitionProperties` is only used to get and compare the transition properties of the recto and verso pages. But these transition properties of course only make sense if they correspond to volume breaks at those pages.

joeha480 replied:
Some notes for my own memory. Commit bb28b93 and a modified version of b13a6dd (using Optional) are ready for merging.

Commit 80185f8 could be implemented by modifying the cost function instead of modifying the calculation provided by the EvenSizeVolumeSplitterCalculator.

joeha480 replied:
I've rewritten aa8bfe9 to avoid using Object and reflection.

joeha480 replied:
Processed b662b0d

joeha480 commented:
Commit bb28b93 and a modified version of b13a6dd (using Optional) are ready for merging.

Commit 80185f8 could be implemented by modifying the cost function instead of modifying the calculation provided by the EvenSizeVolumeSplitterCalculator.

I've rewritten aa8bfe9 to avoid using Object and reflection.

joeha480 commented:
Processed b662b0d

joeha480 commented:
Processed e34b5dd

kalaspuffar commented:
@bertfrees will look at this pull request and resolve conflicts and see if there is any code that could be used still and either fix this PR so it's ready for pulling or creating a new PR with good code.

bertfrees commented:
The following commits were already in master:

- bb28b93: merged in d386eff
- b8dc003: merged in 9ccffb4
- b662b0d: merged in 272e34e
- b13a6dd: merged in d386eff
- e34b5dd: merged in 9ac4e09
- aa8bfe9: merged in 688c1d3
- 00e4bda: merged in 456d999

The following commits have been rebased. If needed they can be split up into multiple PRs.

- 80185f8: rebased to db1e949: This commit should be implemented differently like Joel said above.
- fe7804e: rebased to d9ac9c3
- 051d099: rebased to f747f22: I don't remember exactly why/if this was needed.
- 2678e33: rebased to 640a7bb: Needs some work because it breaks the keep-with-previous-sheets feature.
- 4528bf1: rebased to ee43e16

bertfrees commented:
> Commit 80185f8 could be implemented by modifying the cost function instead of modifying the calculation provided by the EvenSizeVolumeSplitterCalculator.

If I remember correctly what Joel meant by this is that the volume sizes calculated by EvenSizeVolumeSplitter were never meant to be maximum values, only target values.

bertfrees committed:
EvenSizeVolumeSplitter: take into account the extra overhead when adding volumes

bertfrees commented:
> By trying to predict the added overhead for the extra volumes we can improve things.

I have done this now (locally).

bertfrees commented:
I'm going to test some Dedicon books with the new commit (that I haven't pushed yet) and with/without commit db1e949, but I think that's gonna be easier to do when the other PRs have been merged. So focusing on them first.

bertfrees commented:
I rebased the branch onto master and left out one commit (ee43e16) for which I created a [new PR](https://github.com/brailleapps/dotify.formatter.impl/pull/111).

- db1e949 -> a8534a9
- d9ac9c3 -> d6ad13f
- f747f22 -> 6110c1f
- 640a7bb -> 8c184d4
- ee43e16 -> https://github.com/brailleapps/dotify.formatter.impl/pull/111

bertfrees commented:
I have tested all the Dedicon books with only these two commits (and in combination with https://github.com/brailleapps/dotify.formatter.impl/pull/111) and I get no oscillations:
- the "Only save "NextPageDetailsInSequence" and "TransitionProperties" info when needed" commit (d6ad13f) -> rebased to 45ae05b
- and another commit that replaces a8534a9 -> c27f558

"Base 'breakable' info in CrossReferenceHandler on BlockLineLocation" (8c184d4) appears to be not needed (anymore), so I'll leave that one out, at least for now, because it broke the keep-with-previous-sheets feature.

The "Don't remember exactly why this was needed" commit (6110c1f) was probably never needed to fix the oscillation issue, but it is possible that I made the change to fix an issue that I noticed while testing the Dedicon books, or maybe it was just an issue I spotted in the code. Unfortunately I don't remember. I suggest that I move this commit to a separate PR and then we can check how/if this changes the regression test suite, and judge whether it is an improvement or not.

bertfrees committed:
Combine nested if statements

PMD complained about this.

bertfrees commented:
A side effect of c27f558 is still that in some cases the volumes are not as "equal" as before, although it's less the case than with the the original change a8534a9, which consistently allocated too much spaces for volumes. 

The volume-breaks-advanced3-input.obfl test shows that in this particular case the extra overhead for the added volume was overestimated. I don't know if there is a way to avoid this. As I explained before the way the volume breaking works is just very limiting.

bertfrees commented:
@kalaspuffar I think this branch is ready now for running the regression tests on.

kalaspuffar commented:
Hi @bertfrees 

Shouldn't I wait to run the regression test until all the unit tests passes?

Best regards
Daniel

bertfrees commented:
But I explained didn't I? The behavior changed. I can change the expected result of that unit test if you want. The unit test is of course not a very realistic case because the volumes are so tiny. I'm more interested to know what happens with real books.

bertfrees committed:
fixup c27f558: test changed

kalaspuffar commented:
Hi @bertfrees 

I think you should change the test, we won't merge any code that has failing tests. And if you think the test is incomplete or incorrect you may remove it as well. But we need a solid test suite that can find the issues before we find them in production.

Best regards
Daniel

bertfrees commented:
I didn't say anything about merging. Sure, I'll modify the test, no problem. I was just unsure about the result, and I'm trying to get some feedback from you. Firstly, as I said it's difficult to say whether the results of these small tests are fine. We should probably create some tests with more and larger volumes. But I figured we could maybe first look at the books from the regression test suite to get an impression. Secondly, you and I might have different goals. Dedicon's primary goal is to fix the oscillation issue that prevents some books to get through at all. I think that the volume breaking algorithm itself has room for improvements as well, although right now that is of secondary importance to me. On the other side there is MTM who might be more concerned about the continuity in the behavior of the volume breaking and might attach more importance to the "evenness" of volumes. I don't know. So the fact that I am happy with this change doesn't necessarily mean you should be happy with it.

kalaspuffar commented:
Hi @bertfrees 

Great comment about what your goal for this change is. I've now run the regression tests and this change did also change 162 books so I need to look into these further to get a picture of the changes.

Do you have any preference on which result I look at first?

When it comes to expectations and goals, our goal is to progress the work with the product in the most stable manner.

I've not had any discussions with MTM of the specific topic but I don't see any reason why the evenness of volume breaks should be a deal-breaker if we need to improve the code in order to produce some kinds of books.

The main goal for MTM at the moment is to solve the issues Dedicon has with the product so you can produce your workload.

Best regards
Daniel

bertfrees commented:
> Do you have any preference on which result I look at first?

I haven't.

I've been thinking about what is required to validate a test result, when the volume breaks may have moved. There are two parts to it:
- Checking that _only_ the volume breaks have changed and that the actual content is unchanged. Note that it is also possible that page breaks change near the volume breaks. Writing a tool to do this check is an option, but it's not trivial and maybe we are confident enough that it's true by manually checking a few files.
- Checking that the volume breaks are better or equally good as before. I think that it makes little sense to visually check the results and to try to judge whether the volume breaks are good because there are so many things you need to take into account. It's not a simple checklist. It's an equation of several factors. So what about writing a tool that analyzes a PEF and computes a "volume break score" for it, which we can easily compare between different versions? That would probably be the ideal solution, however it is not that simple to write such a tool, mainly because not all required information is present in the PEF. What would be easier to do is compute this score within Dotify because Dotify has all the information it needs to compute it. I realize that this way we basically let the tool evaluate itself, which is not so great, but it's better than nothing I think. Note that it is currently not so that Dotify can simply pick the solution with the optimal score. This is due to the limitations of the current optimization algorithm (greedy approach). A downside of letting Dotify compute the score is that you can only have a single way to compute the score, i.e. there can be only one interpretation of what are good volume breaks (unlike when we would have a tool that would analyze the PEF because we can have several versions of them). But on the other hand if we need multiple interpretations we also need Dotify's volume breaking to become configurable and I'm not sure we should go there.
